sqlite/src/lib.rs

```diff
@@ -126,21 +160,46 @@ impl ConnectionWrapper {
             .await
             .unwrap()?;
         Ok(ConnectionWrapper {
-            conn: Arc::new(Mutex::new(conn)),
+            conn: Arc::new(Mutex::new(Some(conn))),
         })
     }
     /// Interact with the underlying SQLite connection. This function
     /// expects a closure that takes the connection as parameter. The
     /// closure is executed in a separate thread so that the async runtime
     /// is not blocked.
-    pub async fn interact<F, R>(&self, f: F) -> R
+    pub async fn interact<F, R>(&self, f: F) -> Result<R, InteractError>
     where
-        F: FnOnce(&rusqlite::Connection) -> R + Send + 'static,
+        F: FnOnce(&rusqlite::Connection) -> Result<R, rusqlite::Error> + Send + 'static,
         R: Send + 'static,
     {
-        let conn = self.conn.clone();
-        tokio::task::spawn_blocking(move || f(&*conn.lock().unwrap()))
-            .await
-            .unwrap()
+        let arc = self.conn.clone();
+        tokio::task::spawn_blocking(move || {
+            let guard = arc.lock().unwrap();
+            let conn = guard.as_ref().unwrap();
+            f(&conn)
+        })
+        .await
+        .map_err(|e| {
+            if e.is_panic() {
+                InteractError::Panic(e.into_panic())
+            } else if e.is_cancelled() {
+                InteractError::Aborted
+            } else {
+                unreachable!();
+            }
+        })?
+        .map_err(InteractError::Rusqlite)
+    }
+}
+
+impl Drop for ConnectionWrapper {
+    fn drop(&mut self) {
+        let arc = self.conn.clone();
+        // Drop the `rusqlite::Connection` inside a `spawn_blocking`
+        // as the `drop` function of it can block.
+        tokio::task::spawn_blocking(move || match arc.lock() {
+            Ok(mut guard) => guard.take(),
+            Err(e) => e.into_inner().take(),
+        });
```

Comment:
I would add type annotations and explicit `drop`s to make it absolutely clear that you're dropping the connection.

Reply:
Right. `drop(...)` makes it absolutely clear. I'm currently rewriting this as a generic `SyncWrapper` which can be used for `deadpool-diesel` as well.